joy/CMakeLists.txt

```
@@ -1,39 +1,54 @@
 cmake_minimum_required(VERSION 3.5)
-project(joy)
-
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_CXX_STANDARD_REQUIRED ON)
-
-# Load catkin and all dependencies required for this package
-set(CATKIN_DEPS roscpp diagnostic_updater sensor_msgs roslint)
-find_package(catkin REQUIRED ${CATKIN_DEPS})
-
-roslint_cpp()
 
-catkin_package(DEPENDS sensor_msgs)
-
-# Look for <linux/joystick.h>
-include(CheckIncludeFiles)
-check_include_files(linux/joystick.h HAVE_LINUX_JOYSTICK_H)
+project(joy)
 
-if(CATKIN_ENABLE_TESTING)
-  roslint_add_test()
+# Default to C++14
+if(NOT CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 14)
+endif()
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic)
 endif()
 
-if(HAVE_LINUX_JOYSTICK_H)
-  include_directories(msg/cpp ${catkin_INCLUDE_DIRS})
-  add_executable(joy_node src/joy_node.cpp)
-  target_link_libraries(joy_node ${catkin_LIBRARIES})
-
-# Install targets
-  install(TARGETS joy_node
-    ARCHIVE DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
-    LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
-    RUNTIME DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION})
-else()
-  message("Warning: no <linux/joystick.h>; won't build joy node")
+find_package(ament_cmake_ros REQUIRED)
+find_package(rclcpp REQUIRED)
+find_package(rclcpp_components REQUIRED)
+find_package(sdl2_vendor REQUIRED)
+find_package(SDL2 REQUIRED)
+find_package(sensor_msgs REQUIRED)
+
+include_directories(include)
+add_library(joy SHARED src/joy.cpp)
+ament_target_dependencies(joy
+  "rclcpp_components"
+  "rclcpp"
+  "sensor_msgs")
+# Need this since sdl2 doesn't properly export its
+# libraries so ament_target_dependencies can find them
+target_link_libraries(joy
+  SDL2::SDL2)
```

Replace with the following to work with both from-source and from-binary versions:

```suggestion
if(${SDL2_LIBRARIES})
  target_link_libraries(joy
    ${SDL2_LIBRARIES})
elseif(SDL2::SDL2)
  target_link_libraries(joy
    SDL2::SDL2)
endif()
```
After replacing references to `sdl2_vendor` with `sdl2`, the `sdl2_vendor` package can be removed.

---

Tested this with both binary-installed on `Bionic` and source-built.

---

> After replacing references to `sdl2_vendor` with `sdl2`, the `sdl2_vendor` package can be removed.

I don't think we can remove it.  We still need it so that it can build on Windows (which doesn't have a convenient chocolatey package for SDL2).  But I will certainly add the above so that we don't unnecessarily build it on Linux and macOS.

---

Oh, and we still have the problem of what to `find_package` here, since it should be `sdl2` on Linux and `SDL2` everywhere else.

---

This can be accomplished with the following bit of CMake:

```
find_package(SDL2 QUIET)
find_package(sdl2 QUIET)

if(NOT ${SDL2_LIBRARIES} AND NOT SDL2::SDL2)
  message( FATAL_ERROR "SDL2 not found." )
endif()
```

**UNTESTED** (I don't have an easily-accessible, non-Linux environment)

---

I just needed to step away from it for a while.  I finally did manage to figure out how to make this much nicer; now we only build sdl2_vendor at all if we need to, and we only get a single SDL2::SDL2 whether we get it from the "old-style" variables or the "new-style" targets.

---

Maybe I missed this: Why do we still need `sdl2_vendor` instead of using `rosdep`?

---

> Maybe I missed this: Why do we still need `sdl2_vendor` instead of using `rosdep`?

For Windows.  There is no packaged SDL2 library for Windows, so we have to build it there.

---

Right. Well, one of us will have to create a Chocolatey SDL2 package at some point. Guess this works for now.